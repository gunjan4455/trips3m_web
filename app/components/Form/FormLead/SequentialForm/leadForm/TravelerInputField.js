import React from 'react';
import PropTypes from 'prop-types';

class TravelerInputField extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      inputValue: this.props.inputValue
    };
    this.props.dispatchRxFormAction(this.state.inputValue, this.props.fieldName);
  }

  onChange = (e) => {
    let value = parseInt(e.target.value);
    if(this.props.fieldName === 'adult') {
      if (value >= 2) {
        this.setState({ inputValue: value });
      } else {
        value = 1;
        this.setState({ inputValue: 1 });
      }
    }else {
      if(value >=0){
        this.setState({ inputValue: value });
      }else {
        value = 0;
        this.setState({ inputValue: 0 });
      }
    }
    this.props.dispatchRxFormAction(value, this.props.fieldName);
    this.props.handleTravelerCount(value);
  };

  updateDays = (action) => {
    let tempValue = 0;
    let newValue = 0;
    let currentValue = 0;
    if(this.props.fieldName === 'adult'){
      currentValue = this.state.inputValue ? this.state.inputValue : 1;
    }else {
      currentValue = this.state.inputValue ? this.state.inputValue : 0;
    }

    if (action === 'inc') {
      tempValue = 1;
    } else {
      tempValue = -1;
    }
    newValue = parseInt(currentValue, 10) + tempValue;
    if (newValue >= 1 && this.props.fieldName === 'adult') {
      this.setState({ inputValue: newValue });
      this.props.dispatchRxFormAction(newValue.toString(), this.props.fieldName);
    }
    if ((this.props.fieldName === 'infant' || this.props.fieldName === 'child') && newValue >= 0) {
      this.setState({ inputValue: newValue });
      this.props.dispatchRxFormAction(newValue.toString(), this.props.fieldName);
      this.props.handleTravelerCount(newValue);
    }
  };

  handleBlur = (e) => {
    if(e.target.value === '') {
      this.setState({ inputValue: 0 });
    }
  };

  render() {
    return (
        <div >
          <div className="daysCount relative input-with-icon-box input-two-icons input-with-icon-box">
            <button type="button" onClick={() => this.updateDays('dec')}
              className="min-date cursorP minus-icon" />
            <input
              name={this.props.fieldName}
              type="number"
              placeholder={this.props.placeholder}
              className="input text-center duration"
              value={this.state.inputValue}
              onChange={this.onChange}
              id={this.props.fieldName}
              onBlur={this.handleBlur}
            />
            <button type="button" onClick={() => this.updateDays('inc')}
              className="plus-date cursorP plus-icon" />
          </div>
          {
            this.props.meta
            && this.props.meta.touched
            && (this.props.meta.error
            && <span className="input-error-text">{this.props.meta.error}</span>)
          }
        </div>
    );
  }
}

TravelerInputField.propTypes = {
  dispatchRxFormAction: PropTypes.func.isRequired,
  handleTravelerCount: PropTypes.func.isRequired,
  inputValue: PropTypes.string,
  fieldName: PropTypes.string.isRequired,
  meta: PropTypes.object.isRequired,
  formValues: PropTypes.object.isRequired,
  placeholder: PropTypes.string
};

TravelerInputField.defaultProps = {
  inputValue: ''
};

export default TravelerInputField;
