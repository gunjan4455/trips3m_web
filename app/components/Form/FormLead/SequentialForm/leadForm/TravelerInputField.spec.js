import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { shallow, configure } from 'enzyme';

import TravelerInputField from './TravelerInputField';

const mockCallNextSlide = jest.fn();
const mockDispatchRxFormAction = jest.fn();
const mockHandleTravelerCount = jest.fn();
configure({ adapter: new Adapter() });

const DEFAULT_PROPS = {
  callNextSlide: mockCallNextSlide,
  dispatchRxFormAction: mockDispatchRxFormAction,
  input: {
    name: 'TravelerInputField',
    value: '',
    onChange: () => {
    }
  },
  inputValue: 1,
  fieldName: "child",
  formValues: {
    email: null,
    dont_utilize_budget_new: true,
    budget_new: true,
    start_date_noidea: false,
    adult: '2',
    child: '0',
    isd_code: '+91',
    child_age_answers_attributes: [
      {
        user_input: '0',
        customization_id: '7',
        option_id: '29'
      },
      {
        user_input: '0',
        customization_id: '7',
        option_id: '30'
      },
      {
        user_input: '0',
        customization_id: '7',
        option_id: '31'
      }
    ],
    syncErrors: {
      email: 'Enter valid email id'
    },
    to_loc: 'Kerala',
    start_date: '2019-07-16',
    start_date_flexible: false,
    trip_days: '2',
    hotel_needed_answer_attributes: {
      option_id: '95',
      customization_id: '22'
    },
    hotel_category_answers_attributes: [
      {
        option_id: '6',
        customization_id: '2'
      }
    ],
    other_answers_attributes: [
      null,
      {
        option_id: '99',
        customization_id: '24'
      }
    ],
    budget: '8000..8400',
    infant: '0'
  },

  meta: {
    active: false,
    asyncValidating: false,
    autofilled: false,
    dirty: false,
    form: 'LeadForm',
    invalid: false,
    pristine: true,
    submitting: false,
    submitFailed: false,
    touched: false,
    valid: true,
    visited: false
  },

  location: {
    query: {}
  },
  handleTravelerCount: mockHandleTravelerCount

};
const shallowWrapperTravelerInputField = (props) => {
  return shallow(<TravelerInputField {...props} />);
};

describe('lead form module', () => {

  let props = { ...DEFAULT_PROPS };

  afterEach(() => {
    props = { ...DEFAULT_PROPS };
  });

  it('should render correctly', () => {
    const wrapper = shallowWrapperTravelerInputField(props);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call handleBlur properly', () => {
    const e = {
      target: {
        value: ''
      }
    };
    const wrapper = shallowWrapperTravelerInputField(props);
    wrapper.instance().handleBlur(e);
    expect(wrapper.state().inputValue).toEqual(0)
  });

  it('should call onChange properly', () => {
    const e = {
      target: {
        value: 7
      }
    };
    const wrapper = shallowWrapperTravelerInputField(props);
    wrapper.instance().onChange(e);
    expect(wrapper.state().inputValue).toEqual(7);
  });

  it('should call onChange properly when fieldname in props is adult', () => {
    const e = {
      target: {
        value: 1
      }
    };
    props.fieldName = 'adult';
    const wrapper = shallowWrapperTravelerInputField(props);
    wrapper.instance().onChange(e);
    expect(wrapper.state().inputValue).toEqual(1);
  });

  it('should call updateDays properly when updateDays call with inc', () => {
    mockDispatchRxFormAction.mockClear();
    const wrapper = shallowWrapperTravelerInputField(props);
    wrapper.instance().updateDays('inc');
    expect(mockDispatchRxFormAction).toHaveBeenCalledWith('2','child');
  });

  it('should call updateDays properly when updateDays call with other than inc', () => {
    mockDispatchRxFormAction.mockClear();
    const wrapper = shallowWrapperTravelerInputField(props);
    wrapper.instance().updateDays('inc1');
    expect(mockDispatchRxFormAction).toHaveBeenCalledWith('0','child');
  });
  it('should call updateDays properly when updateDays call with inc and fieldname in props is adult', () => {
    mockDispatchRxFormAction.mockClear();
    props.fieldName = 'adult';
    const wrapper = shallowWrapperTravelerInputField(props);
    wrapper.instance().updateDays('inc');
    expect(mockDispatchRxFormAction).toHaveBeenCalledWith('2','adult');
  });

});
