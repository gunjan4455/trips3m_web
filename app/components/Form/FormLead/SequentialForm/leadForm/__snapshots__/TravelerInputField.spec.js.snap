// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`lead form module should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TravelerInputField
    callNextSlide={[MockFunction]}
    dispatchRxFormAction={
      [MockFunction] {
        "calls": Array [
          Array [
            1,
            "child",
          ],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    fieldName="child"
    formValues={
      Object {
        "adult": "2",
        "budget": "8000..8400",
        "budget_new": true,
        "child": "0",
        "child_age_answers_attributes": Array [
          Object {
            "customization_id": "7",
            "option_id": "29",
            "user_input": "0",
          },
          Object {
            "customization_id": "7",
            "option_id": "30",
            "user_input": "0",
          },
          Object {
            "customization_id": "7",
            "option_id": "31",
            "user_input": "0",
          },
        ],
        "dont_utilize_budget_new": true,
        "email": null,
        "hotel_category_answers_attributes": Array [
          Object {
            "customization_id": "2",
            "option_id": "6",
          },
        ],
        "hotel_needed_answer_attributes": Object {
          "customization_id": "22",
          "option_id": "95",
        },
        "infant": "0",
        "isd_code": "+91",
        "other_answers_attributes": Array [
          null,
          Object {
            "customization_id": "24",
            "option_id": "99",
          },
        ],
        "start_date": "2019-07-16",
        "start_date_flexible": false,
        "start_date_noidea": false,
        "syncErrors": Object {
          "email": "Enter valid email id",
        },
        "to_loc": "Kerala",
        "trip_days": "2",
      }
    }
    handleTravelerCount={[MockFunction]}
    input={
      Object {
        "name": "TravelerInputField",
        "onChange": [Function],
        "value": "",
      }
    }
    inputValue={1}
    location={
      Object {
        "query": Object {},
      }
    }
    meta={
      Object {
        "active": false,
        "asyncValidating": false,
        "autofilled": false,
        "dirty": false,
        "form": "LeadForm",
        "invalid": false,
        "pristine": true,
        "submitFailed": false,
        "submitting": false,
        "touched": false,
        "valid": true,
        "visited": false,
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="daysCount relative input-with-icon-box input-two-icons input-with-icon-box"
        >
          <button
            className="min-date cursorP minus-icon"
            onClick={[Function]}
            type="button"
          />
          <input
            className="input text-center duration"
            id="child"
            name="child"
            onBlur={[Function]}
            onChange={[Function]}
            type="number"
            value={1}
          />
          <button
            className="plus-date cursorP plus-icon"
            onClick={[Function]}
            type="button"
          />
        </div>,
        false,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="min-date cursorP minus-icon"
              onClick={[Function]}
              type="button"
            />,
            <input
              className="input text-center duration"
              id="child"
              name="child"
              onBlur={[Function]}
              onChange={[Function]}
              type="number"
              value={1}
            />,
            <button
              className="plus-date cursorP plus-icon"
              onClick={[Function]}
              type="button"
            />,
          ],
          "className": "daysCount relative input-with-icon-box input-two-icons input-with-icon-box",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "min-date cursorP minus-icon",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "input text-center duration",
              "id": "child",
              "name": "child",
              "onBlur": [Function],
              "onChange": [Function],
              "placeholder": undefined,
              "type": "number",
              "value": 1,
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "plus-date cursorP plus-icon",
              "onClick": [Function],
              "type": "button",
            },
            "ref": null,
            "rendered": null,
            "type": "button",
          },
        ],
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="daysCount relative input-with-icon-box input-two-icons input-with-icon-box"
          >
            <button
              className="min-date cursorP minus-icon"
              onClick={[Function]}
              type="button"
            />
            <input
              className="input text-center duration"
              id="child"
              name="child"
              onBlur={[Function]}
              onChange={[Function]}
              type="number"
              value={1}
            />
            <button
              className="plus-date cursorP plus-icon"
              onClick={[Function]}
              type="button"
            />
          </div>,
          false,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="min-date cursorP minus-icon"
                onClick={[Function]}
                type="button"
              />,
              <input
                className="input text-center duration"
                id="child"
                name="child"
                onBlur={[Function]}
                onChange={[Function]}
                type="number"
                value={1}
              />,
              <button
                className="plus-date cursorP plus-icon"
                onClick={[Function]}
                type="button"
              />,
            ],
            "className": "daysCount relative input-with-icon-box input-two-icons input-with-icon-box",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "min-date cursorP minus-icon",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": null,
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "input text-center duration",
                "id": "child",
                "name": "child",
                "onBlur": [Function],
                "onChange": [Function],
                "placeholder": undefined,
                "type": "number",
                "value": 1,
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "plus-date cursorP plus-icon",
                "onClick": [Function],
                "type": "button",
              },
              "ref": null,
              "rendered": null,
              "type": "button",
            },
          ],
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
  Symbol(enzyme.__childContext__): null,
}
`;
